File: Code/Greedy/Couples_Holding_Hands_765.cpp
#include <bits/stdc++.h>
using namespace std;
class Solution
{
    map<int, int> mp;

public:
    int minSwapsCouples(vector<int> row)
    {
        for (int i = 0; i < row.size(); ++i)
        {
            mp[row[i]] = i;
        }
        int first, second, ans = 0;
        for (int i = 0; i < row.size(); i += 2)
        {
            first = row[i];
            if (first & 1)
            {
                if (first - 1 != row[i + 1])
                {
                    ++ans;
                    second = row[i + 1];
                    swap(row[i + 1], row[mp[first - 1]]); //
                    mp[second] = mp[first - 1];
                    mp[first - 1] = i + 1;
                }
            }
            else
            {
                if (first + 1 != row[i + 1])
                {
                    ++ans;
                    second = row[i + 1];
                    swap(row[i + 1], row[mp[first + 1]]); //
                    mp[second] = mp[first + 1];
                    mp[first + 1] = i + 1;
                }
            }
        }
        return ans;
    }
};
int main()
{
#ifndef ONLINE_JUDGE
    freopen("input.txt", "r", stdin);
    freopen("output.txt", "w", stdout);
#endif
    Solution obj;
    auto ans = obj.minSwapsCouples({0, 2, 1, 3});
    // cout << ans;
    return 0;
}

File: java/Greedy_java/Couples_Holding_Hands_765.java
package java.Greedy_java;

import java.util.*;
import java.io.*;

// import static java.lang.Math.max;
// import static java.lang.Math.min;
// import static java.lang.Math.abs;
// import static java.lang.System.out;

public class Couples_Holding_Hands_765 {
    private HashMap<Integer, Integer> mp = new HashMap<>();

    public int minSwapsCouples(int[] row) {
        for (int i = 0; i < row.length; ++i) {
            mp.put(row[i], i);
        }
        int first, second, ans = 0;
        for (int i = 0; i < row.length; i += 2) {
            first = row[i];
            if ((first & 1) == 1) {
                if (first - 1 != row[i + 1]) {
                    ++ans;
                    second = row[i + 1];
                    // swap
                    row[i + 1] ^= row[mp.get(first - 1)];
                    row[mp.get(first - 1)] ^= row[i + 1];
                    row[i + 1] ^= row[mp.get(first - 1)];

                    mp.put(second, mp.get(first - 1));
                    mp.put(first - 1, i + 1);
                }
            } else {
                if (first + 1 != row[i + 1]) {
                    ++ans;
                    second = row[i + 1];
                    // swap
                    row[i + 1] ^= row[mp.get(first + 1)];
                    row[mp.get(first + 1)] ^= row[i + 1];
                    row[i + 1] ^= row[mp.get(first + 1)];

                    mp.put(second, mp.get(first + 1));
                    mp.put(first + 1, i + 1);
                }
            }
        }
        return ans;
    }

    public static void main(String[] args) throws FileNotFoundException {
        File f = new File("output.txt");

        // Scanner ip = new Scanner(System.in);
        try (Scanner ip = new Scanner(new File("input.txt"));
                PrintWriter op = new PrintWriter(f);) {
            Couples_Holding_Hands_765 obj = new Couples_Holding_Hands_765();
            int grid[] = { 0, 2, 1, 3 };
            op.println(obj.minSwapsCouples(grid));

        }
    }

}
